Add Portis web3 provider via a wrapper

jessgusclark opened the pull request:
Adds support for Portis as a web3 provider.

It also adds a loading wheel after the user selects a provider. This is also shown when the user is connecting the DataVault. i.e. User clicks on Metamask, the waiting wheel is shown until the user is connected.

## Details

Portis is not an EIP1193 provider, so after Portis is passed to rLogin, a wrapper is created that allows EIP1193 methods to be called using `request`. The tested methods are:

- `eth_chainId`
- `eth_accounts`
- `net_version`
- `personal_sign` - Portis requires that the message be encoded as HEX, which this wrapper does for the user. I'm unsure if this is "standard" (before 1193) or not, so right now this wrapper will only be used with Portis.
- `eth_sendTransaction`

## Apps tested with:

- [x] RIF Id Manager
- [x] Email VC Issuer
- [x] Sample App with all 3 variants: `dapp`, `open`, `permissioned`

jessgusclark committed:
Add Portis as an example to the sample.

jessgusclark committed:
Create wrapper, and get it to return chain_id, accounts, and sign simple messages.

jessgusclark committed:
Rename to portisWrapper and write tests.

jessgusclark committed:
Support net_version and eth_sendTransaction

- The error is in the response, not the err prop. Reject the correct message

jessgusclark committed:
Add logout functionality and refactor params.

jessgusclark committed:
Add loading wheel when waiting for the DV or the provider.

jessgusclark committed:
Use testnet and remove console.log

jessgusclark committed:
Update portisWrapper and index based on id manager integration.

jessgusclark committed:
Remove the useEffect thing.

jessgusclark committed:
Fix lint and test issues.

ilanolkies commented on src/Core.tsx:
```diff
@@ -280,6 +283,11 @@ export class Core extends React.Component<IModalProps, IModalState> {
 
+    // portis
+    if (provider.isPortis) {
+      provider._portis.logout()
```
We should await this, it is a `Promise` - https://github.com/portis-project/web-sdk/blob/develop/packages/portis-web3/src/index.ts#L126

should we move this to a `disconnectPortis` method? As for WC

ilanolkies replied:
I see now WC `disconnect` returns a `Promise` too
https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/ethereum-provider/src/index.ts#L116

jessgusclark replied:
Great! I have added a `disconnect` method to the wrapper. This mirrors how walletconnect works. So, now when disconnecting, I can check if this method exists and call it.

ilanolkies commented on src/lib/portisWrapper.ts:
```diff
@@ -0,0 +1,33 @@
+  }
+}
+
+export const portisWrapper = (provider: any) => new Proxy(provider, handler)
```
Excellent! Some minor things:
- Should we add support for `eth_signTypedData` and `eth_sendTransaction` too? I don’t know if they work already
- Can we list the methods we tested?
- Let’s add some comments about this code

```suggestion
// This wrapper makes Portis Web SDK EIP-1193 compatible
export const portisWrapper = (provider: any) => new Proxy(provider, handler)
```

jessgusclark replied:
I have added additional comments to the wrapper and added a section in the readme for Portis. Regarding `eth_signTypeData` it does not work out of the box as Portis' implementation is a bit different than Metamask's. Should I take the time now to implement this?

ilanolkies reviewed (commented):
Nice onee!

ilanolkies commented on src/Core.tsx:
```diff
@@ -316,11 +324,12 @@ export class Core extends React.Component<IModalProps, IModalState> {
       {currentStep === 'Step3' && <ConfirmSelectiveDisclosure did={(chainId && address) ? did : ''} sd={sd!} onConfirm={this.onConfirmAuth} />}
       {currentStep === 'error' && <ErrorMessage title={errorReason?.title} description={errorReason?.description}/>}
       {currentStep === 'wrongNetwork' && <WrongNetworkComponent supportedNetworks={supportedChains} isMetamask={isMetamask(provider)} changeNetwork={this.changeMetamaskNetwork} />}
+      {currentStep === 'loading' && <Loading size={10} color="#008FF7" />}
```
Cool

jessgusclark committed:
Add disconnect() method to the wrapper

- This mirror's how WalletConnect disconnects and makes it easier to call the single function when it exists.

jessgusclark committed:
Update readme to add section on Portis.

jessgusclark committed:
Add eth_signTypedData v1 as an example to the demo.

jessgusclark committed:
return logout as a function and catch if the provider throws an error.

jessgusclark committed:
Update readme to include gasPrice estimation issue.

jessgusclark committed:
Refactor sendingTransaction to get the gasPrice if not set by the developer.

jessgusclark committed:
Modify test to return a fake gas value when asked, write additional test for the gasPrice.

jessgusclark committed:
Update readme portis description.

jessgusclark committed:
Get the gas price for eth_sendRawTransaction as well.

jessgusclark commented:
Rebased.

ilanolkies commented on src/lib/portisWrapper.ts:
```diff
@@ -0,0 +1,54 @@
+            if ((method === 'eth_sendTransaction' || method === 'eth_sendRawTransaction') && !params[0].gasPrice) {
+              return target.send({ method: 'eth_getBlockByNumber', params: ['latest', false] },
+                (err: string | null, res: any) =>
+                  err ? reject(res.err) : sendTransaction(method, [{ ...params[0], gasPrice: res.result.minimumGasPrice }])
```
```suggestion
                  err ? reject(res.err) : sendTransaction(method, [{ ...params[0], gasPrice: res.result.minimumGasPrice * 1.01 }])
```
Or we can set even a little more

> Nodes that forward transactions could check that the advertised gas price in a transaction is at least 1% higher than the minimum. This assures the transaction a lifetime of 100 blocks assuming a constantly increasing block minGasPrice.  Ref: https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP09.md

jessgusclark replied:
Added.

ilanolkies commented:
Rebased version bumping

ilanolkies committed:
Bump versoin

ilanolkies commented:
Let's add the last comment and merge the beta version into develop. After we give it a test we can bump and release major.

jessgusclark committed:
Get gas price if not set and increase it by 1%.

jessgusclark committed:
Update README to explain developers with Portis and send.

jessgusclark commented:
After talking with @miltonwg , I have updated the description in the Readme to indicate that the wrapper will add the `gasPrice` when sending `request`. However, the developer will need to add a `gasPrice` if they use `send` or `sendAsync`.

I also bumped the `gasPrice` by 1% as requested by @ilanolkies . This PR is ready to go, below is the status of the App's support of Portis. I checked the Apps below using `yarn link` as a few extra commits have been added.

- [x] VC Issuer - Worked fine as it does not have any transactions, only signatures. Portis support is already merged into main. [deployed to production here](https://email-verifier.identity.rifos.org/)
- [x] RIF Id Manager - As Ken found, the identity section does not work. After talking with Milton, it was decided to show a message letting the user know they can't update their data. [A PR is ready to be merged here](https://github.com/rsksmart/rif-identity-manager/pull/81). 
- [x] RNS Manager React - Works locally, however, the published rLogin package 1.0.7-beta.1 does not support rawTransactions, so [this PR](https://github.com/rnsdomains/rns-manager-react/pull/417) will need to wait until the next version is released then updated.

---
update: this comment was originally posted on the RIF Identity Manager PR.

ilanolkies commented:
Do we add it to v1.0.8? In that case let;s merge it before #96

ilanolkies committed:
Merge branch 'develop' into portis-wrapper
